src/App.tsx

```diff
@@ -1,26 +1,192 @@
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [errorMessage, setErrorMessage] = useState<string>('');
+  const [filter, setFilter] = useState<FilterStatus>(FilterStatus.All);
+  const [todoOnLoading, setTodoOnLoading] = useState<number[] | null>(null);
+  const [tempTodo, setTempTodo] = useState<null | Todo>(null);
+
+  const loadTodos = () => {
+    if (!USER_ID) {
+      return;
+    }
+
+    setIsLoading(true);
+    setErrorMessage('');
+
+    getTodos()
+      .then(result => setTodos(result))
+      .catch(err => setErrorMessage(err.message || 'Unable to load todos'))
+      .finally(() => setIsLoading(false));
+  };
+
+  useEffect(() => {
+    loadTodos();
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const removeTodo = (todoId: number) => {
+    setTodoOnLoading([todoId]);
+    deleteTodos(todoId)
+      .then(() =>
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => todo.id !== todoId),
+        ),
+      )
+      .catch(err => setErrorMessage(err.message || 'Unable to delete a todo'))
+      .finally(() => setTodoOnLoading(null));
+  };
+
+  const handleDeleteCompleted = () => {
+    const completedTodosId = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    setTodoOnLoading(completedTodosId);
+
+    Promise.allSettled(completedTodosId.map(id => deleteTodos(id)))
+      .then(results => {
+        const successfulIds = completedTodosId.filter(
+          (_, index) => results[index].status === 'fulfilled',
+        );
+
+        const hasErrors = results.some(result => result.status === 'rejected');
+
+        if (hasErrors) {
+          setErrorMessage('Unable to delete a todo');
+        }
+
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => !successfulIds.includes(todo.id)),
+        );
+      })
+      .catch(err => setErrorMessage(err.message || 'Unable to delete todos'))
+      .finally(() => setTodoOnLoading(null));
+  };
+
+  const handleAddTodos = async (title: string) => {
+    let hasError = false;
+
+    if (title.trim().length === 0) {
+      setErrorMessage('Title should not be empty');
+
+      return null;
+    }
+
+    const todo: Omit<Todo, 'id'> = {
+      completed: false,
+      title: title.trim(),
+      userId: USER_ID,
+    };
+
+    setTempTodo({ ...todo, id: 0 });
+    try {
+      const newTodo = await addTodo(todo);
+
+      setTodos(prevTodos => [...prevTodos, newTodo]);
+    } catch (error: any) {
+      setErrorMessage(error.message || 'Unable to add a todo');
+      hasError = true;
+    } finally {
+      setTempTodo(null);
+    }
+
+    return hasError;
+  };
+
+  const filteredTodos = todos.filter(todo => {
+    switch (filter) {
+      case FilterStatus.All:
+        return true;
+      case FilterStatus.Active:
+        return !todo.completed;
+      case FilterStatus.Completed:
+        return todo.completed;
+      default:
+        return true;
```

Comment:
you can combine cases
```suggestion
    switch (filter) {
      case FilterStatus.Active:
        return !todo.completed;
      case FilterStatus.Completed:
        return todo.completed;
       case FilterStatus.All:
      default:
        return true;
```